File: gomoku/gomoku-haskell/src/Gomoku.hs
{-# LANGUAGE NamedFieldPuns #-}

module Gomoku
  ( playGame,
    createGame,
  )
where

import Data.Maybe
import System.Console.ANSI
import System.IO (BufferMode (NoBuffering), hReady, hSetBuffering, hSetEcho, stdin)

data Stone = StoneO | StoneX | Blank | Highlight deriving (Eq)

instance Show Stone where
  show StoneO = "○"
  show StoneX = "●"
  show Blank = " "
  show Highlight = "." -- Current selection

data Board = Board
  { grid :: [Stone], -- A list to track the placement of stones
    width :: Int -- An integer to store the dimension of the grid (width of the side)
  }

instance Show Board where
  show Board {grid, width}
    | gridLength == width ^ 2 = horizontalEdge ++ "\n" ++ "║" ++ showFirst ++ "|" ++ showRest
    | mod gridLength width == 0 = "║" ++ showFirst ++ "|" ++ showRest
    | gridLength == 1 = showFirst ++ "║\n" ++ horizontalEdge
    | mod gridLength width == 1 = showFirst ++ "║\n" ++ showRest
    | otherwise = showFirst ++ "|" ++ showRest
    where
      showFirst = show $ head grid
      showRest = show $ Board {grid = tail grid, width = width}
      gridLength = length grid
      horizontalEdge = replicate (width * 2 + 1) '='

createBoard :: Int -> Board
createBoard width = Board {grid = replicate (width ^ 2) Blank, width = width}

updateBoard :: Stone -> Int -> Board -> Maybe Board
updateBoard stone pos board@Board {grid, width}
  | stoneAtPos == StoneO || stoneAtPos == StoneX = Nothing
  | otherwise = Just $ Board {grid = replacedGrid, width = width}
  where
    stoneAtPos = grid !! pos
    replacedGrid = replaceNth pos stone grid
    replaceNth :: Int -> a -> [a] -> [a]
    replaceNth n newVal xs = x ++ (newVal : ys)
      where
        (x, _ : ys) = splitAt n xs

data Player = PlayerA | PlayerB deriving (Eq, Show)

playerToStone :: Player -> Stone
playerToStone PlayerA = StoneX
playerToStone PlayerB = StoneO

data Game = Game
  { player :: Player, -- Current player
    board :: Board, -- Board state
    selection :: Int, -- Current selection
    stoneCount :: Int -- Number of stones being placed on the board
  }

data Direction = Up | Right | Down | Left deriving (Eq)

nextSelection :: Direction -> Int -> Int -> Int
nextSelection direction width selection = case direction of
  Gomoku.Up -> finalizeSelection $ selection - width
  Gomoku.Down -> finalizeSelection $ selection + width
  Gomoku.Right -> finalizeSelection $ selection + 1
  Gomoku.Left -> finalizeSelection $ selection - 1
  where
    total = width ^ 2
    finalizeSelection x = mod (x + total) total -- Mod the selection index with the size of board to make sure
    -- the index does not overflow

createGame :: Int -> Game
createGame w = Game {player = PlayerA, board = initializedBoard, selection = selection, stoneCount = 0}
  where
    selection = w ^ 2 `div` 2 -- Try to select the center of the board
    freshBoard = createBoard w
    initializedBoard = fromMaybe freshBoard $ updateBoard Highlight selection freshBoard

instance Show Game where
  show game = concat ["\n", show $ board game, "\n\n", "Turn: ", show $ player game, "\n"]

goToNextAvailPos :: Direction -> Game -> Game
goToNextAvailPos direction game@Game {player, board, selection, stoneCount} = goToNextAvailPos' direction selection game
  where
    goToNextAvailPos' :: Direction -> Int -> Game -> Game
    goToNextAvailPos' direction initPos game@Game {player, board, selection, stoneCount}
      | nSelection == initPos = game
      | otherwise = case updatedBoard of
        Nothing -> goToNextAvailPos' direction initPos game {board = clearedBoard, selection = nSelection}
        Just b -> game {board = b, selection = nSelection}
      where
        clearedBoard = fromMaybe board $ updateBoard Blank selection board
        nSelection = nextSelection direction (width board) selection
        updatedBoard = updateBoard Highlight nSelection clearedBoard

placeStone :: Game -> Game
placeStone game@Game {player, board, selection, stoneCount} =
  case updatedBoard of
    Just b -> game {board = b, stoneCount = stoneCount + 1, player = nextPlayer player}
    Nothing -> game
  where
    updatedBoard = updateBoard (playerToStone player) selection board
    nextPlayer PlayerA = PlayerB
    nextPlayer PlayerB = PlayerA

updateGame :: String -> Game -> Game
updateGame "\ESC[A" game = goToNextAvailPos Gomoku.Up game
updateGame "\ESC[B" game = goToNextAvailPos Gomoku.Down game
updateGame "\ESC[C" game = goToNextAvailPos Gomoku.Right game
updateGame "\ESC[D" game = goToNextAvailPos Gomoku.Left game
updateGame "\n" game = placeStone game
updateGame _ game = game

checkWin :: Game -> Bool
checkWin game@Game {player, board, selection, stoneCount} = any (\d -> extension (x, y) d >= 5) dirs
  where
    dirs = [(0, 1), (1, 0), (1, 1), (1, -1)] -- Four directions (vertical, horizontal, and two diagonals) need to be checked
    w = width board
    x = selection `div` w
    y = selection `mod` w
    curGrid = grid board
    curStone = curGrid !! selection
    validPos x y = x >= 0 && y >= 0 && x < w && y < w
    extensionInDir (x, y) (dX, dY)
      | validPos x y && curGrid !! (x * w + y) == curStone =
        1 + extensionInDir (x + dX, y + dY) (dX, dY)
      | otherwise = 0
    extension (x, y) (dX, dY) =
      extensionInDir (x, y) (dX, dY) + extensionInDir (x, y) (- dX, - dY) - 1

checkDraw :: Game -> Bool
checkDraw game@Game {player, board, selection, stoneCount} = stoneCount == width board ^ 2

getKey :: IO String
getKey = reverse <$> getKey' ""
  where
    getKey' chars = do
      char <- getChar
      more <- hReady stdin
      (if more then getKey' else return) (char : chars)

playGame :: Game -> IO ()
playGame game = do
  clearScreen
  playGame' game
  where
    playGame' game = do
      hSetBuffering stdin NoBuffering
      hSetEcho stdin False
      printGame game
      input <- getKey
      let newGame = updateGame input game
      if input == "\n" && checkWin newGame
        then do
          printGame newGame
          putStrLn $ show (player game) ++ " won!"
        else
          if checkDraw newGame
            then do
              printGame newGame
              putStrLn "It is a draw!"
            else playGame' newGame
      where
        printGame game = do
          cursorUpLine numLines
          putStr gameStr
          where
            gameStr = show game
            numLines = length . filter ('\n' ==) $ gameStr

File: gomoku/gomoku-python/gomoku.py
import sys
import termios
import tty
from typing import List, Optional
from enum import Enum, auto
from contextlib import contextmanager

class Stone(Enum):
    STONE_O = "○"
    STONE_X = "●"
    BLANK = " "
    HIGHLIGHT = "."  # Current selection

    def __str__(self):
        return self.value

class Board:
    def __init__(self, grid: List[Stone], width: int):
        self.grid = grid
        self.width = width

    def __str__(self):
        grid_length = len(self.grid)
        horizontal_edge = (self.width * 2 + 1) * "="
        if grid_length == self.width ** 2:
            return horizontal_edge + "\n" + "║" + str(self.grid[0]) + "|" + str(Board(self.grid[1:], self.width))
        elif grid_length % self.width == 0:
            return "║" + str(self.grid[0]) + "|" + str(Board(self.grid[1:], self.width))
        elif grid_length == 1:
            return str(self.grid[0]) + "║\n" + horizontal_edge
        elif grid_length % self.width == 1:
            return str(self.grid[0]) + "║\n" + str(Board(self.grid[1:], self.width))
        else:
            return str(self.grid[0]) + "|" + str(Board(self.grid[1:], self.width))

def create_board(width: int) -> Board:
    return Board([Stone.BLANK] * (width ** 2), width)

def update_board(stone: Stone, pos: int, board: Board) -> Optional[Board]:
    if board.grid[pos] in [Stone.STONE_O, Stone.STONE_X]:
        return None
    return Board(board.grid[:pos] + [stone] + board.grid[pos + 1:], board.width)

class Player(Enum):
    PLAYER_A = auto()
    PLAYER_B = auto()

    def __str__(self) -> str:
        if self == Player.PLAYER_A:
            return "Player A"
        elif self == Player.PLAYER_B:
            return "Player B"

def player_to_stone(player: Player) -> Stone:
    return Stone.STONE_X if player == Player.PLAYER_A else Stone.STONE_O

class Direction(Enum):
    UP = auto()
    RIGHT = auto()
    DOWN = auto()
    LEFT = auto()

def next_selection(direction: Direction, width: int, selection: int) -> int:
    total = width ** 2
    new_selection = {
        Direction.UP: selection - width,
        Direction.DOWN: selection + width,
        Direction.RIGHT: selection + 1,
        Direction.LEFT: selection - 1,
    }[direction]
    return (new_selection + total) % total

class Game:
    def __init__(self, player: Player, board: Board, selection: int, stone_count: int):
        self.player = player
        self.board = board
        self.selection = selection
        self.stone_count = stone_count

    def __str__(self):
        return f"\n{self.board}\n\nTurn: {self.player}\n"

def create_game(width: int) -> Game:
    selection = width ** 2 // 2
    fresh_board = create_board(width)
    initialized_board = update_board(Stone.HIGHLIGHT, selection, fresh_board)
    return Game(Player.PLAYER_A, initialized_board, selection, 0)

def go_to_next_avail_pos(direction: Direction, game: Game) -> Game:
    def go_to_next_avail_pos_helper(direction: Direction, init_pos: int, game: Game) -> Game:
        cleared_board = update_board(Stone.BLANK, game.selection, game.board)
        if cleared_board is None:
            cleared_board = game.board
        new_selection = next_selection(direction, game.board.width, game.selection)
        updated_board = update_board(Stone.HIGHLIGHT, new_selection, cleared_board)

        if new_selection == init_pos:
            return game
        elif updated_board is None:
            return go_to_next_avail_pos_helper(direction, init_pos, Game(game.player, cleared_board, new_selection, game.stone_count))
        else:
            return Game(game.player, updated_board, new_selection, game.stone_count)

    return go_to_next_avail_pos_helper(direction, game.selection, game)

def place_stone(game: Game) -> Game:
    updated_board = update_board(player_to_stone(game.player), game.selection, game.board)
    next_player = Player.PLAYER_B if game.player == Player.PLAYER_A else Player.PLAYER_A

    if updated_board is None:
        return game
    else:
        return Game(next_player, updated_board, game.selection, game.stone_count + 1)

def update_game(input_str: str, game: Game) -> Game:
    if input_str == "\x1b[A":
        return go_to_next_avail_pos(Direction.UP, game)
    elif input_str == "\x1b[B":
        return go_to_next_avail_pos(Direction.DOWN, game)
    elif input_str == "\x1b[C":
        return go_to_next_avail_pos(Direction.RIGHT, game)
    elif input_str == "\x1b[D":
        return go_to_next_avail_pos(Direction.LEFT, game)
    elif input_str in ("\n", "\r"):
        return place_stone(game)
    else:
        return game
    

def check_win(game: Game) -> bool:
    def extension(x: int, y: int, dx: int, dy: int) -> int:
        return extension_in_dir(x, y, dx, dy) + extension_in_dir(x, y, -dx, -dy) - 1

    def extension_in_dir(x: int, y: int, dx: int, dy: int) -> int:
        if valid_pos(x, y) and game.board.grid[x * game.board.width + y] == cur_stone:
            return 1 + extension_in_dir(x + dx, y + dy, dx, dy)
        else:
            return 0

    def valid_pos(x: int, y: int) -> bool:
        return 0 <= x < game.board.width and 0 <= y < game.board.width

    dirs = [(0, 1), (1, 0), (1, 1), (1, -1)]  # Vertical, horizontal, and two diagonals to be checked
    width = game.board.width
    selection = game.selection
    x, y = selection // width, selection % width
    cur_stone = game.board.grid[selection]
    return any(extension(x, y, dx, dy) >= 5 for dx, dy in dirs)

def check_draw(game: Game) -> bool:
    return game.stone_count == game.board.width ** 2

@contextmanager
def terminal_settings(file_descriptor: int):
    old_settings = termios.tcgetattr(file_descriptor)
    try:
        tty.setraw(file_descriptor)
        yield
    finally:
        termios.tcsetattr(file_descriptor, termios.TCSADRAIN, old_settings)

def get_key() -> str:
    with terminal_settings(sys.stdin.fileno()):
        ch = sys.stdin.read(1)
        if ch == "\x1b":
            ch += sys.stdin.read(2)
        return ch

def clear_screen() -> None:
    print("\033[H\033[J", end="")

def play_game(game: Game) -> None:
    clear_screen()
    play_game_helper(game)

def play_game_helper(game: Game) -> None:
    print(game, end="")
    input_str = get_key()
    new_game = update_game(input_str, game)

    clear_screen()

    if check_win(new_game):
        print(new_game)
        print(f"{game.player} won!")
    elif check_draw(new_game):
        print(new_game)
        print("It is a draw!")
    else:
        play_game_helper(new_game)

def main() -> None:
    game = create_game(15)
    play_game(game)


if __name__ == "__main__":
    main()